### PR TITLE
[ALLUXIO-2160] Write unit tests for AlluxioMasterRestServiceHandler

### PR DESCRIPTION
**Pull request by schernolyas**
https://alluxio.atlassian.net/browse/ALLUXIO-2160

### CODE REVIEW

**Comment by schernolyas** [comment]
Hi @aaudiber !

I found solution. The problem was in PowerMock.
Please  recheck!

**Comment by schernolyas** [comment]
Hi @aaudiber !
Problem tests are:
1) alluxio.master.lineage.LineageMasterTest
2) alluxio.worker.block.SpaceReserverTest

Main problem is PowerMock uses other classpath for loading mocked classes. In this way, library jre/lib/amd64/libjaas_unix.so try to load twice  (and get java.lang.UnsatisfiedLinkError)

Can I redevelop these tests?
Who can consult me?

**Comment by schernolyas** [comment]
Hi @aaudiber 

I see http://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html. And ...  We have only one way for fix the problem:  restart JVM for earch test. 
Reason: PowerMock use own classpath and alluxio.security.LoginUser  get exception then it try to load native library again.

**Comment by aaudiber** [comment]
@schernolyas Thanks for digging into this! Is the issue with `Powermock` independent of this PR? Can you explain how it leads to the errors along the lines of 
```
java.lang.AssertionError: Entry must be a Long! expected:<class java.lang.Long> but was:<class java.lang.String>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at alluxio.master.AlluxioMasterRestServiceHandlerTest.getUfsCapacityBytes(AlluxioMasterRestServiceHandlerTest.java:350)
```
?

**Comment by schernolyas** [comment]
Hi @aaudiber !

PowerMock uses different classloader for loading mocks of final classes.  If You try to load native library. at second time .then You get the exception. Native library 'jre/lib/amd64/libjaas_unix.so' loading occurs in class  alluxio.security.LoginUser (see javax.security.auth.login.LoginContext). In this way .... We get a choice:
1) not mock final classes (and redevelop these classes)
2) reload JVM after executing each test class

I selected point 2.
For details  You can see to method 'loadLibrary0' of class java.lang.ClassLoader

File: core/server/pom.xml [hunk 1]
```diff
@@ -249,6 +249,20 @@
...
+      <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-surefire-plugin</artifactId>
+          <version>2.19</version>
+          <configuration>
+            <argLine>-Djava.net.preferIPv4Stack=true -XX:MaxPermSize=512M</argLine>
+            <redirectTestOutputToFile>${test.output.redirect}</redirectTestOutputToFile>
+            <!-- Alphabetical order on even hours, reverse-alphabetical on odd hours  -->
+            <runOrder>hourly</runOrder>
+            <reuseForks>false</reuseForks>
```
Comment (aaudiber) [inline core/server/pom.xml]: This change is very significant - could we move it to its own PR?

File: core/server/pom.xml [hunk 2]
```diff
@@ -249,6 +249,20 @@
...
+            <forkCount>2</forkCount>
```
Comment (aaudiber) [inline core/server/pom.xml]: Could we make this "1C" instead so that it uses one fork per core on the machine running the build?

File: core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+  private ServletContext mContext;
+  private BlockMaster mBlockMaster;
+  private AlluxioMasterRestServiceHandler mHandler;
+  private ManualClock mClock;
+  private ExecutorService mExecutorService;
+  @Rule
+  public TemporaryFolder mTestFolder = new TemporaryFolder();
+
+  @BeforeClass
+  public static void setUpClass() throws Exception {
```
Comment (aaudiber) [inline core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java]: nit: we usually call this method `beforeClass`

File: core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    String filesPinnedProperty =
+        MetricsSystem.getMasterMetricName(FileSystemMaster.Metrics.FILES_PINNED);
+    MetricsSystem.METRIC_REGISTRY.remove(filesPinnedProperty);
+  }
+
+  @Before
+  public void setUp() throws Exception {
```
Comment (aaudiber) [inline core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java]: nit: we usually call this method `before`

File: core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+    when(underFileSystemMock.getSpace(TEST_PATH, UnderFileSystem.SpaceType.SPACE_TOTAL))
+        .thenReturn(UFS_SPACE_TOTAL);
+    when(underFileSystemMock.getSpace(TEST_PATH, UnderFileSystem.SpaceType.SPACE_USED)).thenReturn(
+        UFS_SPACE_USED);
+    when(underFileSystemFactoryMock.create(eq(TEST_PATH), anyObject())).thenReturn(
+        underFileSystemMock);
+    UnderFileSystemRegistry.register(underFileSystemFactoryMock);
+  }
+
+  @After
```
Comment (aaudiber) [inline core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java]: nit: we usually call this method `after`

Comment (schernolyas) [inline core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java]: fixed

File: core/server/pom.xml (hunk 2 above)
Comment (schernolyas) [inline core/server/pom.xml]: Hi @aaudiber !
No.
____________________________________
Tests run: 45, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.949 sec <<< FAILURE! - in alluxio.master.file.PermissionCheckTest
setStateFail(alluxio.master.file.PermissionCheckTest)  Time elapsed: 0.087 sec  <<< ERROR!
alluxio.exception.AccessControlException: Permission denied: user=admin, access=-w-, path=/testDir: failed at /
        at alluxio.master.file.PermissionCheckTest.createDirAndFileForTest(PermissionCheckTest.java:177)
        at alluxio.master.file.PermissionCheckTest.before(PermissionCheckTest.java:158)
_____________________________________________________

File: core/server/pom.xml (hunk 1 above)
Comment (schernolyas) [inline core/server/pom.xml]: Hi @aaudiber !

I don't know. You can try.  But .... the PR is not works without the change.

File: core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java (hunk 5 above)
Comment (schernolyas) [inline core/server/src/test/java/alluxio/master/AlluxioMasterRestServiceHandlerTest.java]: fixed

File: core/server/pom.xml (hunk 2 above)
Comment (schernolyas) [inline core/server/pom.xml]: No.
________________________________________________
Tests run: 45, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.949 sec <<< FAILURE! - in alluxio.master.file.PermissionCheckTest
setStateFail(alluxio.master.file.PermissionCheckTest)  Time elapsed: 0.087 sec  <<< ERROR!
alluxio.exception.AccessControlException: Permission denied: user=admin, access=-w-, path=/testDir: failed at /
        at alluxio.master.file.PermissionCheckTest.createDirAndFileForTest(PermissionCheckTest.java:177)
        at alluxio.master.file.PermissionCheckTest.before(PermissionCheckTest.java:158)
__________________________________________

File: core/server/pom.xml (hunk 1 above)
Comment (schernolyas) [inline core/server/pom.xml]: I think no. Because the PR not works without the change..

**Comment by schernolyas** [comment]
Hi @aaudiber !

Recheck the PR please

File: core/server/pom.xml (hunk 1 above)
Comment (aaudiber) [inline core/server/pom.xml]: I'm making the change here: https://github.com/Alluxio/alluxio/pull/4291

Comment (schernolyas) [inline core/server/pom.xml]: Hi @aaudiber !
What is our next steps?

**Comment by schernolyas** [comment]
Hi @aaudiber !
What is our next steps?

**Comment by aaudiber** [comment]
@schernolyas Now that https://github.com/Alluxio/alluxio/pull/4291 has merged, can you merge master into your PR?

**Comment by schernolyas** [comment]
Hi @aaudiber !
I merged upstream/master to the branch.  Recheck please

File: core/server/pom.xml [hunk 6]
```diff
@@ -249,6 +249,20 @@
           <!-- insert other configuration here (optional) -->
         </configuration>
       </plugin>
+      <plugin>
```
Comment (aaudiber) [inline core/server/pom.xml]: This change has been added to the root pom file, so we don't need to modify `core/server/pom.xml` any more

**Comment by schernolyas** [comment]
Hi @aaudiber !

Recheck please!

File: core/server/pom.xml [hunk 7]
```diff
@@ -248,7 +248,7 @@
           </analysis>
           <!-- insert other configuration here (optional) -->
         </configuration>
-      </plugin>
+      </plugin>
```
Comment (aaudiber) [inline core/server/pom.xml]: Can you revert this added whitespace? After that we should be good to merge

**Comment by schernolyas** [comment]
Hi @aaudiber !
Spaces removed. Recheck please.

**Comment by schernolyas** [comment]
Hi @aaudiber !
Okey. Thank you very much!
